activitysim/core/fast_eval.py

```
@@ -0,0 +1,98 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+import pandas as pd
+from pandas import eval as _eval
+
+if TYPE_CHECKING:
+    from collections.abc import Hashable, Iterator, Mapping, Sequence
+
+    from pandas._typing import ArrayLike
+
+
+def _get_cleaned_column_resolvers(
+    df: pd.DataFrame, raw: bool = True
+) -> dict[Hashable, ArrayLike | pd.Series]:
+    """
+    Return the special character free column resolvers of a dataframe.
+
+    Column names with special characters are 'cleaned up' so that they can
+    be referred to by backtick quoting.
+    Used in :meth:`DataFrame.eval`.
+    """
+    from pandas import Series
+    from pandas.core.computation.parsing import clean_column_name
+
+    if isinstance(df, pd.Series):
+        return {clean_column_name(df.name): df}
+
+    # CHANGED FROM PANDAS: do not even convert the arrays to pd.Series, just
+    # give the raw arrays to the compute engine. This is potentially a breaking
+    # change if any of the operations in the eval string require a pd.Series.
```

This means we cannot have a utility expression specific to pd.Series, such as .map(). Could we create a check on user specified expressions to alert/avoid this? Could this be done as part of https://github.com/ActivitySim/activitysim/pull/936? And the alternative is to move strings require pd.Series to preprocessors?

---

Proactively catching every utility expression specific to pd.Series is tricky, as we'd need to parse the expression and analyze it, and we'd need something like a comprehensive list of code features that work with pd.Series but not a simple array.

The solution for now is down on line 87, in the try-except:  first we try to eval the string assuming it will work with just arrays, and if it fails we retry with the make-it-series solution.  This relies on the assumption that eval expressions that depend on the inputs being pd.Series will raise an exception when run on arrays, and not return values that are different from expected.  In my experience the exception-raising is much more likely, but the latter is perhaps possible. I have found and fixed similar problems when comparing legacy-vs-sharrow code -- which is a slightly different kind of problem but in a very similar vein, and extremely challenging to diagnose. Only a very careful and rigorous testing of models will ever uncover this kind of thing.

---

Thank you for the thoughtful answer. Yea, failing silently sounds like a tough challenge. I don't know if I can come up with a better solution than what you've implemented. Perhaps before the try-except block, scan the expression string for know pd.Series features (acknowledge it won't be a comprehensive list) and force it go to the make-it-series solution, so that it reduces the reliance on the exception?

---

I think parsing these expressions trying to catch these rare problems is going to add an undesirable amount of runtime overhead to every component, on every model run.  On the flip side, I think it's a better solution to expect model designers will test their specifications and check they are running correctly.  One useful tool to do this testing is by running with sharrow in test mode, which completely circumvents this problem.